### PR TITLE
Offline Mention Emails

### PR DESCRIPTION
**Pull request by TheConnMan**
The beginnings of offline mention emails.

Things I need help with:

- [x] i18n translations
- [x] Confirmation that I didn't totally screw preference persistence up
- [x] Comments on if pulling back user `emails` and `settings` from the DB on every mention/DM is too much and if so a better solution
- [x] Comments on if an email for every mention and DM is a useful feature by itself or if there has to be message grouping so 10 DMs in a row don't drive people nuts (maybe they should?)

Ref #741

### CODE REVIEW

**Comment by marceloschmidt** [comment]
That's looking good! I think user should have options: every message, daily digest, disabled (and maybe others). Instead of using a boolean setting, I'd use a `select`, as in `Accounts_RegistrationForm`.

File: packages/rocketchat-lib/server/functions/sendMessage.coffee [hunk 1]
```diff
@@ -79,6 +79,13 @@ RocketChat.sendMessage = (user, message, room, options) ->
 						query:
 							userId: userOfMention._id
 
+				if userOfMention.status is 'offline' and userOfMention.settings?.preferences?.emailNotificationMode is true
+					Email.send
+						to: userOfMention.emails[0].address
```
Comment (marceloschmidt) [inline packages/rocketchat-lib/server/functions/sendMessage.coffee]: There are cases where userOfMention doesn't have an e-mail address on file (such as OAuth login). A few OAuths also don't provide an e-mail address, such as Facebook.

Comment (TheConnMan) [inline packages/rocketchat-lib/server/functions/sendMessage.coffee]: Do you know off-hand if `userOfMention.emails` is `undefined` or `[]` if a user has no emails? I'll check it in the if statement.

**Comment by TheConnMan** [comment]
I definitely agree that a daily digest option would be nice. If there is a cron job I can hook into an an easy way to filter all the daily messages I can add it in, though I don't know that I'm the best person to do so. I'm happy do dive in, I just don't want the email on every message functionality to get held up because I'm having to pick up Meteor.

Is it worthwhile to add in the plain boolean with the added option as an additional feature later? We could even change the boolean to a select with only two options in anticipation for the later functionality.

File: packages/rocketchat-lib/server/functions/sendMessage.coffee (hunk 1 above)
Comment (marceloschmidt) [inline packages/rocketchat-lib/server/functions/sendMessage.coffee]: AFAIK, it's undefined, but check for both cases, please.

**Comment by marceloschmidt** [comment]
I agree, make it a select for now, with an option for `disabled` and another for `every message`. This way, we can improve your work and do it for daily digests. We already use a `cron` system (`percolate:synced-cron`), which you can check in `server/startup/cron.coffee`, just in case you want to touch that area ;) If you don't, just let us know and we'll merge your PR and continue from there. Thank you very much for your interest.

**Comment by TheConnMan** [comment]
The two changes mentioned above have been added in. I hope to get to the daily digest functionality at some point, but it won't be for at least a week. How should i18n translations be dealt with? After the merge by people who know those languages?

**Comment by engelgabriel** [comment]
Hi @TheConnMan awesome new feature! Thanks for the PR!

If you can just pull from our develop and resolve the conflicts, I can accept the PR.

The i18n translations are done by the community using LingoHub.

**Comment by TheConnMan** [comment]
@engelgabriel Happy to help. I just merged in the changes and there was some significant merging to be done. I'd appreciate a second set of eyes to make sure everything went smoothly.

**Comment by engelgabriel** [comment]
Thanks @TheConnMan 
@sampaiodiego and @marceloschmidt can you please review this one? I'd like to merge it asap.

**Comment by sampaiodiego** [comment]
should this send an e-mail to each user of the room if a `@all` is used?

**Comment by sampaiodiego** [comment]
I have a few considerations on this. Just checking everyone's opinion:

- we don't have to send an email on `@all` mentions - not by default at least;
- we have to check the `statusConnection` field (not your fault @TheConnMan =) there is no docs about this field) because a user with _invisible_ status is is considered _offline_;
- should we send emails just to verified addresses or try to send to any address?

found a bug when you set your desktop notification preferences to **ALL** messages you got an email to every message in that channel. we need to separate the this like I did in #2198 to avoid this kind of problem.. I'm able to do it after this got merged :wink: 

@RocketChat/core thoughts? 

this is a very useful feature, thanks again @TheConnMan :smile:

**Comment by engelgabriel** [comment]
I will accept this PR as a partial solution to #741 now, as I want to merge it asap to avoid greater conflicts with https://github.com/RocketChat/Rocket.Chat/pull/2198

Some comments on what needs to be done still:
- We should not send an email on @all mentions by default, but must be configurable.
- We have to check the statusConnection field.
- We must send emails just to verified addresses.

Also important:
- We would like @jd365 opinion here, as he is one of the main backers of the issue.
- I agree with @jclabaugh about "mimicking Slack on this feature" as much as possible.

**Pull request merged**